cmds/switch_root/switch_root.go

```diff
@@ -23,140 +24,201 @@ func usage() string {
+
+		for _, name := range names {
+			// try to open it with O_DIRECTORY to know if it's a directory
+			// NOTE O_DIRECTORY is linux specific
+			namefd, err := syscall.Openat(fd, name, syscall.O_DIRECTORY, syscall.O_RDWR)
+			if err != nil {
+				// just delete files
+				if err := unlinkat(fd, name, 0); err != nil {
+					return err
+				}
+			} else {
+				defer syscall.Close(namefd)
+
+				var nameStatT syscall.Stat_t
+				if err := fstatat(fd, name, &nameStatT, unix.AT_SYMLINK_NOFOLLOW); err != nil {
+					return err
+				}
+
+				if fdDev, err := getDev(fd); err != nil {
+					return err
+				} else if fdDev != nameStatT.Dev {
+					// on different filesystem, don't go there
+					continue
+				}
+
+				// not actually a dir, but a symlink to a dir, treat like file and remove
+				if isSymlink(nameStatT.Mode) {
+					if err := unlinkat(fd, name, 0); err != nil {
+						return err
+					}
+					continue
+				}
+
+				// actual recursion
+				if err := recursiveDelete(namefd); err != nil {
+					return err
+				}
+				// back from recursion, dir is not empty, delete
+				if err := unlinkat(fd, name, AT_REMOVEDIR); err != nil {
+					return err
+				}
+			}
 		}
 	}
-
 	return nil
 }
 
-// execCommand will run the executable at "path" with PID 1.
+// do a proper exec to retain PID 1.
 // It returns an error if the command exits incorrectly.
 func execCommand(path string) error {
-	cmd := exec.Command(path)
-	cmd.Stdin = os.Stdin
-	cmd.Stderr = os.Stderr
-	cmd.Stdout = os.Stdout
-	cmd.SysProcAttr = &syscall.SysProcAttr{
-		Setctty:    true,
-		Setsid:     true,
-		Cloneflags: syscall.CLONE_THREAD | syscall.CLONE_NEWPID,
-	}
-
-	return cmd.Run()
+	return syscall.Exec(path, []string{path}, []string{})
 }
 
-// specialFS creates and mounts proc, sys and dev at the root level.
-func specialFS() error {
-
-	if err := syscall.Mkdir("/proc", 0); err != nil {
-		return err
+// check if dir is empty
+func isEmpty(name string) (bool, error) {
+	f, err := os.Open(name)
+	if err != nil {
+		return false, err
 	}
+	defer f.Close()
 
-	if err := syscall.Mkdir("/sys", 0); err != nil {
-		return err
+	_, err = f.Readdirnames(1)
+	if err == io.EOF { // this means emtpy
+		return true, nil
 	}
+	return false, err
+}
 
-	if err := syscall.Mkdir("/dev", 0); err != nil {
-		return err
+// check if file exists
+func isExists(name string) (bool, error) {
+	if _, err := os.Stat(name); err != nil {
+		if os.IsNotExist(err) {
+			return false, nil
+		}
+		return false, err
 	}
+	return true, nil
+}
 
-	if err := syscall.Mount("", "/proc", "proc", 0, ""); err != nil {
-		return err
-	}
+// moves a mountpoint
+func moveMount(oldPath string, newPath string) error {
+	return syscall.Mount(oldPath, newPath, "", syscall.MS_MOVE, "")
+}
 
-	if err := syscall.Mount("", "/sys", "sysfs", 0, ""); err != nil {
-		return err
-	}
+// specialFS creates and mounts proc, sys, and dev.
+func specialFS(newRoot string) error {
+	var mounts = []string{"/dev", "/proc", "/sys", "/run"}
```

Comment:
Is there a reference point for what other projects do for this? I'm a bit skeptical of just having hard-coded constants here. 

I see three options:

1. keep this as is, maybe expand the list a bit
2. rely on new root's init to mount file systems they want where they want them
3. iterate through /proc/mounts and move every fs that isn't /

Reply:
So this comes from the "reference implementation" in `util-linux`, see here: https://git.kernel.org/pub/scm/utils/util-linux/util-linux.git/tree/sys-utils/switch_root.c#n130

We can obviously make it more clever. Let me fixup all the other things, then we can revisit this.